compat.c

```diff
@@ -176,59 +657,1456 @@ ssize_t iio_scan_context_get_info_list(struct iio_scan_context *ctx,
+			return (ssize_t) err;
+
+		compat->enqueued = true;
+		compat->curr = (compat->curr + 1) % compat->nb_blocks;
+	}
+
+	if (!compat->enabled) {
+		err = IIO_CALL(iio_buffer_enable)(buf);
+		if (err)
+			return err;
+
+		compat->enabled = true;
+	}
+
+	compat->all_enqueued |= compat->curr == 0;
+
+	if (!compat->cyclic) {
+		if (compat->all_enqueued) {
+			err = IIO_CALL(iio_block_dequeue)(compat->blocks[compat->curr],
+							  compat->nonblock);
+			if (err < 0)
+				return (ssize_t) err;
+		}
+
+		compat->enqueued = false;
+	}
+
+	return size;
+}
+
+ssize_t iio_buffer_push(struct iio_buffer *buf)
+{
+	struct iio_buffer_compat *compat = IIO_CALL(iio_buffer_get_data)(buf);
+
+	return iio_buffer_push_partial(buf, compat->samples_count);
+}
+
+void * iio_buffer_start(const struct iio_buffer *buf)
+{
+	struct iio_buffer_compat *compat = IIO_CALL(iio_buffer_get_data)(buf);
+	struct iio_block *block = compat->blocks[compat->curr];
+
+	return IIO_CALL(iio_block_start)(block);
+}
+
+void * iio_buffer_first(const struct iio_buffer *buf,
+			const struct iio_channel *chn)
+{
+	struct iio_buffer_compat *compat = IIO_CALL(iio_buffer_get_data)(buf);
+	struct iio_block *block = compat->blocks[compat->curr];
+
+	return IIO_CALL(iio_block_first)(block, chn);
+}
+
+ptrdiff_t iio_buffer_step(const struct iio_buffer *buf)
+{
+	const struct iio_device *dev = IIO_CALL(iio_buffer_get_device)(buf);
+	const struct iio_channels_mask *mask;
+
+	mask = IIO_CALL(iio_buffer_get_channels_mask)(buf);
+
+	return IIO_CALL(iio_device_get_sample_size)(dev, mask);
+}
+
+void * iio_buffer_end(const struct iio_buffer *buf)
+{
+	struct iio_buffer_compat *compat = IIO_CALL(iio_buffer_get_data)(buf);
+	struct iio_block *block = compat->blocks[compat->curr];
+
+	return IIO_CALL(iio_block_end)(block);
+}
+
+ssize_t iio_buffer_foreach_sample(struct iio_buffer *buf,
+				  ssize_t (*callback)(const struct iio_channel *chn,
+						      void *src, size_t bytes,
+						      void *d),
+				  void *data)
+{
+	const struct iio_device *dev = IIO_CALL(iio_buffer_get_device)(buf);
+	struct iio_buffer_compat *compat = IIO_CALL(iio_buffer_get_data)(buf);
+	struct iio_device_compat *dev_compat = IIO_CALL(iio_device_get_data)(dev);
+	struct iio_block *block = compat->blocks[compat->curr];
+
+	return IIO_CALL(iio_block_foreach_sample)(block,
+						  dev_compat->mask,
+						  callback, data);
+}
+
+static void compat_lib_init(void)
+{
+	struct compat *ctx;
+	int err = -ENOMEM;
+
+	ctx = malloc(sizeof(*ctx));
+	if (!ctx) {
+		fprintf(stderr, "Unable to alloc compat structure\n");
+		goto err_set_errno;
+	}
+
+	ctx->lib = iio_dlopen(LIBIIO1_NAME);
+	if (!ctx->lib) {
+		fprintf(stderr, "Unable to dlopen " LIBIIO1_NAME "\n");
+		err = -ENOENT;
+		goto err_free_ctx;
+	}
+
+#define FIND_SYMBOL(lib, symbol) \
+	do { \
+		ctx->symbol = iio_dlsym(lib, #symbol); \
+		if (!ctx->symbol) { \
+			fprintf(stderr, "Unable to find symbol: %s\n", #symbol); \
+			err = -ENOENT; \
+			goto err_dlclose; \
+		} \
```

Comment:
I honestly do not know if we can ever get any strange issue but according to the `dlsym()` man page, the proper way to handle errors is:

```
dlerror(); 
ctx->symbol = iio_dlsym(lib, #symbol);
if (dlerror()) {
     /* error handling */
}

```

Apparently in some weird cases it's actually expected that `dlysm()` returns NULL

Reply:
My man page says `On failure, they return NULL`.

Reply:
Nevermind, I overlooked it.

Anyway... I honestly think we can ignore this case. I can assure it will never happen. Programs are loaded at address 0x0 on Linux, unless requested otherwise during the linking phase with e.g. `-Wl,-Ttext-segment=0x10000000`. In both cases, the run-time linker will load the shared libraries at a higher address; so I do not see any possible case where we get a symbol at address 0x0.

Reply:
I'll just paste the man page

``` text
There are several scenarios when the address of a global symbol is NULL.  For example, a symbol can be placed at
zero address by the linker, via a linker script or with --defsym command-line option.  Undefined weak symbols also have NULL value.  Finally, the symbol value may be the result of a GNU  indirect  function  (IFUNC) resolver function that
returns NULL as the resolved value.  In the latter case, dlsym() also returns NULL without error.  However, in the former
two cases, the behavior of GNU dynamic linker is inconsistent: relocation processing succeeds and the symbol can be
observed to have NULL value, but dlsym() fails and dlerror() indicates a lookup error.
```
I guess the above is likely not to happen with us but OTOH, the code for handling this properly is not *so* horrible so I would probably play safe in here.

That said, is really up to you... I won't be pushing for this